src/python/pants/backend/python/target_types_rules.py

```diff
@@ -38,36 +38,84 @@
 # -----------------------------------------------------------------------------------------------
 
 
-@rule
+@rule(desc="Determining the entry point for a `pex_binary` target")
 async def resolve_pex_entry_point(request: ResolvePexEntryPointRequest) -> ResolvedPexEntryPoint:
-    entry_point_value = request.entry_point_field.value
-    if entry_point_value and not entry_point_value.startswith(":"):
-        if entry_point_value in ("<none>", "<None>"):
-            return ResolvedPexEntryPoint(None)
-        return ResolvedPexEntryPoint(entry_point_value)
-    binary_source_paths = await Get(
-        Paths, PathGlobs, request.sources.path_globs(FilesNotFoundBehavior.error)
-    )
-    if len(binary_source_paths.files) != 1:
-        instructions_url = "https://www.pantsbuild.org/docs/python-package-goal#creating-a-pex-file-from-a-pex_binary-target"
-        if not entry_point_value:
-            raise InvalidFieldException(
-                "Both the `entry_point` and `sources` fields are not set for the target "
-                f"{request.sources.address}, so Pants cannot determine an entry point. Please "
-                "either explicitly set the `entry_point` field and/or the `sources` field to "
-                "exactly one file. You can set `entry_point='<none>' to leave off the entry point."
-                f"See {instructions_url}."
-            )
-        else:
+    ep_val = request.entry_point_field.value
+    ep_alias = request.entry_point_field.alias
+    address = request.entry_point_field.address
+    # This code is tricky, as we support several different schemes:
+    #  1) `<none>` or `<None>` => set to `None`.
+    #  2) `path.to.module` => preserve exactly.
+    #  3) `path.to.module:func` => preserve exactly.
+    #  4) `app.py` => convert into `path.to.app`.
+    #  5) `app.py:func` => convert into `path.to.app:func`.
+    #  6) `:func` => if there's a sources field, convert to `path.to.sources:func` (soon to be deprecated).
+    #  7) no entry point field, but `sources` field => convert to `path.to.sources` (soon to be deprecated).
```

Comment:
After these cases are deprecated, might it be possible to share this logic with the AWS impl?

Reply:
Yes, it will be. I'll add a comment to that effect.